servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/NettyHttpServer.java

```diff
@@ -413,16 +412,17 @@ public void onComplete() {
                         }));
 
                 if (drainRequestPayloadBody) {
-                    return responseWrite.concat(defer(() -> (payloadSubscribed.get() ?
-                            // Discarding the request payload body is an operation which should not impact the state of
-                            // request/response processing. It's appropriate to recover from any error here.
-                            // ST may introduce RejectedSubscribeError if user already consumed the request payload body
-                            requestCompletion : request.messageBody().ignoreElements().onErrorComplete())
-                            // No need to make a copy of the context in both cases.
-                            .shareContextOnSubscribe()));
-                } else {
-                    return responseWrite.concat(requestCompletion);
+                    responseWrite = responseWrite.afterFinally(() -> {
+                        if (!payloadSubscribed.get()) {
+                            // At this point, discarding the request payload body is an operation which should not
+                            // impact the state of request/response processing, and we need to do it in order to both
+                            // reuse the connection and finish any operators that may depend on both the request and
+                            // response bodies to have been fully processed.
+                            request.messageBody().ignoreElements().shareContextOnSubscribe().subscribe();
```

Comment:
This will cover onComplete and onError cases well. In case of the cancellation, shouldn't cancel propagate up to the `connection.read()` to send the `RST_STREAM` for h2 or close the connection for h1?

Reply:
Generally speaking, cancellation is probably the better behavior for draining: I don't know why we don't always subscribe-cancel and if the underlying resource really wants to try and preserve itself (such as a HTTP/1.x session that wants to avoid connection churn) that's its business.

Reply:
This is a bigger discussion and will require more refactoring of the current flow. We can still achieve the expected behavior here if we use `TerminalSignalConsumer` instead of a runnable for `afterFinally` to handle different callbacks differently (there is `CancelImmediatelySubscriber` to help). 

Another way is to reuse the `Single.concatPropagateCancel(Publisher)` operator that we already use in this class in `handleResponse` for a similar purpose. It will just require a bit more conversions from `Completable` to `Single` and then back to `Completable` at the end because we don't have a similar variant for `Completable`.